### PR TITLE
Warn if modifier uses msg.value in non-payable function

### PR DESCRIPTION
**Pull request by erak**
Closes https://github.com/ethereum/solidity/issues/3770

### CODE REVIEW

**Comment by erak** [comment]
solc-js tests are failing. Working on the fix...

File: libsolidity/analysis/ViewPureChecker.cpp [hunk 1]
```diff
@@ -226,19 +226,25 @@ void ViewPureChecker::reportMutability(StateMutability _mutability, SourceLocati
...
 				"Function declared as " +
 				stateMutabilityToString(m_currentFunction->stateMutability()) +
 				", but this expression (potentially) modifies the state and thus "
 				"requires non-payable (the default) or payable."
 			);
+			m_errors = true;
+		}
 		else if (_mutability == StateMutability::Payable)
 		{
 			if (m_modifierInvocation)
```
Comment (axic) [inline libsolidity/analysis/ViewPureChecker.cpp]: Shouldn't this be elevated from warning to error?

File: libsolidity/analysis/ViewPureChecker.cpp [hunk 2]
```diff
@@ -237,6 +239,15 @@ void ViewPureChecker::reportMutability(StateMutability _mutability, SourceLocati
 				", but this expression (potentially) modifies the state and thus "
 				"requires non-payable (the default) or payable."
 			);
+		else if (_mutability == StateMutability::Payable)
+		{
+			if (m_modifierInvocation)
+				m_errorReporter.warning(
```
Comment (axic) [inline libsolidity/analysis/ViewPureChecker.cpp]: Shouldn't this be elevated from warning to error?

Comment (erak) [reply libsolidity/analysis/ViewPureChecker.cpp]: That's a good question. Using ``msg.value`` in a non-payable function also results in a warning: ``"msg.value" used in non-payable function. Do you want to add the "payable" modifier to this function?"``
I'd either elevate both from warning to error or none of them. As far as I understand, this is a security feature that doesn't introduce undefined behaviour (please correct me if I wrong here :wink:), so we could also just keep the warning.

Comment (axic) [reply libsolidity/analysis/ViewPureChecker.cpp]: Ah ok, if the other case is a warning, this PR shouldn't change that, but we should discuss that separately and make a decision.

File: test/libsolidity/syntaxTests/virtualLookup/modifiers_in_libraries.sol [hunk 3]
```diff
@@ -12,3 +12,4 @@ contract Test {
     }
 }
 // ----
+// Warning: (52-61): Modifier used in non-payable function, but this expression requires the function to be payable.
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/virtualLookup/modifiers_in_libraries.sol]: Library functions cannot be payable, so we either have to just exclude them or extend the information to all functions that use the library function.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/401_does_not_warn_msg_value_in_modifier_following_non_payable_public_function.sol [hunk 4]
```diff
@@ -1,4 +0,0 @@
-contract c {
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/401_does_not_warn_msg_value_in_modifier_following_non_payable_public_function.sol]: Why did you delete this test?

File: test/libsolidity/syntaxTests/memberLookup/msg_value_modifier_pure.sol [hunk 5]
```diff
@@ -0,0 +1,6 @@
+contract C {
+    modifier costs(uint _amount) { require(msg.value >= _amount); _; }
+    function f() costs(1 ether) public pure {}
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/memberLookup/msg_value_modifier_pure.sol:3]: Can you add another test where the modifier does not use `msg.value`, but it is invoked as `function f() mod(msg.value) public view {}`?

File: libsolidity/analysis/ViewPureChecker.h [hunk 6]
```diff
@@ -30,6 +30,10 @@ namespace dev
 {
 namespace solidity
 {
+struct MutabilityLocation {
```
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.h]: `{` on a line of its own

File: libsolidity/analysis/ViewPureChecker.h [hunk 7]
```diff
@@ -59,21 +63,24 @@ class ViewPureChecker: private ASTConstVisitor
...
 	/// Creates appropriate warnings and errors and sets @a m_currentBestMutability.
-	void reportMutability(StateMutability _mutability, SourceLocation const& _location);
+	void reportMutability(StateMutability _mutability, SourceLocation const& _location, SourceLocation const& _nestedLocation = SourceLocation());
 
 	std::vector<std::shared_ptr<ASTNode>> const& m_ast;
 	ErrorReporter& m_errorReporter;
 
 	bool m_errors = false;
-	StateMutability m_currentBestMutability = StateMutability::Payable;
+	MutabilityLocation m_bestMutabilityLocation = MutabilityLocation{StateMutability::Payable, SourceLocation()};
```
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.h]: Judging just from the name, I would think this is only a location. Perhaps `m_currentBestMutabilityAndLocation`? (also rename the struct itself)

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/ViewPureChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/ViewPureChecker.cpp]: Tracked in #4601.

File: libsolidity/analysis/ViewPureChecker.cpp [hunk 8]
```diff
@@ -332,15 +365,22 @@ void ViewPureChecker::endVisit(IndexAccess const& _indexAccess)
 	}
 }
 
+bool ViewPureChecker::visit(ModifierInvocation const&)
+{
+	m_modifierInvocation = true;
```
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.cpp]: Is this really needed?

File: libsolidity/analysis/ViewPureChecker.cpp [hunk 9]
```diff
@@ -219,36 +221,51 @@ void ViewPureChecker::endVisit(InlineAssembly const& _inlineAssembly)
...
 				"Function declared as " +
 				stateMutabilityToString(m_currentFunction->stateMutability()) +
 				", but this expression (potentially) modifies the state and thus "
 				"requires non-payable (the default) or payable."
 			);
+			m_errors = true;
+		}
+		else if (_mutability == StateMutability::Payable)
+		{
+			if (m_modifierInvocation)
```
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.cpp]: It might be better to just turn this into a function parameter.

**Comment by chriseth** [comment]
I'm still not too happy about this change.

Also shouldn't we move the msg.value check from bool StaticAnalyzer::visit(MemberAccess const& _memberAccess) to here?

File: test/libsolidity/syntaxTests/memberLookup/msg_value_modifier_pure.sol (hunk 5 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/memberLookup/msg_value_modifier_pure.sol:3]: Added the test.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
This needs a second review.

**Comment by leonardoalt** [comment]
Didn't https://github.com/ethereum/solidity/issues/4601#issuecomment-410356565 decide that it should be an error?

**Comment by chriseth** [comment]
Let me make it an error and let's see what the external tests say.

**Comment by chriseth** [comment]
The way around this error would be to create another `internal` function that would not be forced to be payable, right?

**Comment by leonardoalt** [comment]
I think so, yes

**Comment by leonardoalt** [comment]
Tests fail after turning the warning into error:
- [X] End to end tests - fixed here by https://github.com/ethereum/solidity/pull/4590/commits/1657c6d0b9917e2961d9764e81b14d69858575b1
- [X] External tests - fine
- [X] SolcJS tests - fixed by https://github.com/ethereum/solc-js/pull/244

File: Changelog.md [hunk 10]
```diff
@@ -87,6 +87,7 @@ Compiler Features:
  * Tests: Determine transaction status during IPC calls.
  * Code Generator: Allocate and free local variables according to their scope.
  * Removed ``pragma experimental "v0.5.0";``.
+ * View Pure Checker: Warn about ``msg.value`` used by modifier in non-payable function.
```
Comment (chriseth) [inline Changelog.md]: Perhaps `Warn about a modifier that contains ``msg.value`` being used in a non-payable function`?

Comment (chriseth) [reply Changelog.md]: or 'applied to a non-payable function'?

Comment (chriseth) [reply Changelog.md]: Ah, and this is not a compiler feature - I think it should be a breaking change.

Comment (chriseth) [reply Changelog.md]: Oh, did I write that? :)

**Comment by chriseth** [comment]
Do we need to squash this, @axic?

**Pull request merged**